File: sdks/go/cmd/elydora/plugins/cline.go
package plugins

import "fmt"

// ClinePlugin manages Cline's native global file hooks.
type ClinePlugin struct {
	rename renameFunc
}

// ManagesGuardRuntime reports that Cline commits both generated runtimes and
// both native hook files in one transaction.
func (p *ClinePlugin) ManagesGuardRuntime() bool {
	return true
}

// PreflightInstall validates hook ownership and runtime identity before writes.
func (p *ClinePlugin) PreflightInstall(config InstallConfig) error {
	paths, guard, audit, err := readClineHookPair()
	if err != nil {
		return err
	}
	if err := requireAvailableClineHook(guard); err != nil {
		return err
	}
	if err := requireAvailableClineHook(audit); err != nil {
		return err
	}
	_, err = preflightClineInstallation(config, paths)
	return err
}

func (p *ClinePlugin) Install(config InstallConfig) error {
	paths, guard, audit, err := readClineHookPair()
	if err != nil {
		return err
	}
	if err := requireAvailableClineHook(guard); err != nil {
		return err
	}
	if err := requireAvailableClineHook(audit); err != nil {
		return err
	}
	runtimePaths, err := preflightClineInstallation(config, paths)
	if err != nil {
		return err
	}
	changes, err := prepareClineInstallationChanges(
		config,
		runtimePaths,
		guard,
		audit,
	)
	if err != nil {
		return err
	}
	if err := writeClineChanges(
		changes,
		"Install Cline hooks",
		p.rename,
		runtimePaths,
	); err != nil {
		return err
	}
	fmt.Println("Cline: user-level PreToolUse and PostToolUse hooks installed.")
	return nil
}

func (p *ClinePlugin) Uninstall(agentID string) error {
	paths, guard, audit, err := readClineHookPair()
	if err != nil {
		return err
	}
	changes, err := prepareClineUninstallChanges(
		[]clineHookFile{guard, audit},
		agentID,
	)
	if err != nil {
		return err
	}
	return writeClineHookChanges(
		changes,
		"Uninstall Cline hooks",
		p.rename,
		paths,
	)
}

func (p *ClinePlugin) Status() (PluginStatus, error) {
	paths, guard, audit, err := readClineHookPair()
	status := PluginStatus{
		AgentName:   clineAgentKey,
		DisplayName: "Cline",
		ConfigPath:  paths.hooksDirectory,
	}
	if err != nil {
		return status, err
	}
	contract, err := clineContractForFiles(guard, audit)
	if err != nil || contract == nil {
		return status, err
	}
	status.HookConfigured = true
	status.HookScriptExists, err = clineRuntimeFilesExist(contract)
	if err != nil {
		return status, err
	}
	status.Installed = status.HookScriptExists
	return status, nil
}

File: sdks/node/src/plugins/cline.ts
import type { AgentPlugin, InstallConfig, PluginStatus } from './base.js';
import { AGENT_KEY, resolveHookFiles, runtimeContract } from './cline-contract.js';
import {
  commitClineInstallation,
  commitClineUninstall,
  preflightClineInstallation,
  prepareClineInstallation,
  prepareClineUninstall,
} from './cline-installation.js';
import { readHookFile, requireAvailableHookFile, runtimeFilesExist } from './cline-io.js';
import { SUPPORTED_AGENTS } from './registry.js';

const entry = SUPPORTED_AGENTS.get(AGENT_KEY)!;

async function readHookPair() {
  const paths = resolveHookFiles();
  const [guardFile, auditFile] = await Promise.all([
    readHookFile(paths.guardPath),
    readHookFile(paths.auditPath),
  ]);
  return { paths, guardFile, auditFile };
}

function requireAvailablePair(
  guardFile: Awaited<ReturnType<typeof readHookFile>>,
  auditFile: Awaited<ReturnType<typeof readHookFile>>,
): void {
  requireAvailableHookFile(guardFile);
  requireAvailableHookFile(auditFile);
}

export const clinePlugin: AgentPlugin = {
  managesRuntime: true,

  async preflightInstall(config: InstallConfig): Promise<void> {
    const { guardFile, auditFile } = await readHookPair();
    requireAvailablePair(guardFile, auditFile);
    await preflightClineInstallation(config, [guardFile, auditFile]);
  },

  async install(config: InstallConfig): Promise<void> {
    const { guardFile, auditFile } = await readHookPair();
    requireAvailablePair(guardFile, auditFile);
    const prepared = await prepareClineInstallation(config, guardFile, auditFile);
    await commitClineInstallation(prepared);
    console.log('  Cline: user-level PreToolUse and PostToolUse hooks installed.');
  },

  async uninstall(agentId?: string): Promise<void> {
    const { guardFile, auditFile } = await readHookPair();
    const prepared = await prepareClineUninstall([guardFile, auditFile], agentId);
    await commitClineUninstall(prepared);
  },

  async status(): Promise<PluginStatus> {
    const { paths, guardFile, auditFile } = await readHookPair();
    const contract = runtimeContract(guardFile, auditFile);
    const hookConfigured = contract !== undefined;
    const hookScriptExists = contract ? await runtimeFilesExist(contract) : false;
    return {
      installed: hookConfigured && hookScriptExists,
      agentName: AGENT_KEY,
      displayName: entry.name,
      hookConfigured,
      hookScriptExists,
      configPath: paths.hooksDirectory,
    };
  },
};
